### DISCUSSION TITLE
using onErrorRetry without compromising on Exponential backoff algorithm

### DISCUSSION THREAD

**Discussion by sairajchouhan** [vercel/swr#1574] [General]
### Essentially I want to do these
- Do not retry if the status of the response is `404`
- Retry for all other errors

 I am using  `onErrorRetry` in global `SWRConfig` for this purpose.
```tsx
// _app.tsx
 <SWRConfig
      value={{
        fetcher: (baseURL, config: AxiosRequestConfig) => axios({ ...config, baseURL }),
        onErrorRetry: (err) => {
          if (err.response.status === 404) return
        },
      }}
    >
      <Component {...pageProps} />
</SWRConfig>
```
 but it seems that I do not retry exponentially after I provide `onErrorRetry`, and it is well documented too
> You can also override this behavior via the `onErrorRetry` option:

Is there any way, that I can stop retrying for `404` errors without compromising on the Exponential backoff algorithm for other types of errors

Click  [here ](https://codesandbox.io/s/swr-on-error-retry-g79xm?file=/pages/index.tsx) for the codesandbox with demonstration

**Comment by mpeterdev** [comment]
I think this behavior would be a great addition to the library. Would the team accept a contribution to add a configuration which allows bailing on certain errors without overriding the entire error retry process?

I think it would make sense to be able to provide a function that receives the error and returns a boolean on whether to proceed with the default retry logic. I could see this being a new config or updating `shouldRetryOnError` to also accept a function

https://github.com/vercel/swr/blob/de17eb6ce925ae73548f8b7ec5dfd46d91f2b3c8/src/use-swr.ts#L307

could be updated to something like

```ts
if ((typeof config.shouldRetryOnError === 'function' && config.shouldRetryOnError(err)) || (typeof config.shouldRetryOnError === 'boolean' && config.shouldRetryOnError)) {
```

then if you want to keep using exponential backoff but bail only on 401s for example, you could provide the following

```ts
shouldRetryOnError: (err) => err.statusCode !== 401
```

**Reply by shuding** [reply]
> updating shouldRetryOnError to also accept a function

This sounds like a great feature to have, feel free to open a PR for it! 👍

**Reply by sairajchouhan** [reply]
I thought this was not a good thing to have since no one replied to it😅

**Comment by zgreen** [comment]
Was looking for the same thing and stumbled on this discussion. But it seems this feature was actually added in 1.2.1: https://github.com/vercel/swr/releases/tag/1.2.1 (it's not just not well documented right now)

So, thanks!

**Comment by wmadden** [comment]
For anyone who finds this thread in the future, you can simply delegate to the original SWR implementation of `onErrorRetry()`, which implements the exponential backoff, from your own implementation. E.g.:

```tsx
import { defaultConfig } from "swr/_internal";

<SwrConfig value={{
  onErrorRetry(error, ...args) {
    if ( ... ) {
      // Your custom logic
    } else {   
      // Otherwise use the default behavior
      defaultConfig.onErrorRetry(error, ...args);
    }
  }}
/>
```

**Reply by samuliasmala** [reply]
Has anyone found a good solution for typings when delegating the `onErrorRetry()`? The type of the `config` parameter is not correct when delegating it:

![image](https://user-images.githubusercontent.com/14218719/219021565-174a9a53-f154-4c4d-83a0-469d5c39e7f6.png)

The error message I get for the underlined `config` is as follows:
```
Argument of type 'Readonly<PublicConfiguration<ResData, ErrorResponse, (args: string) => FetcherResponse<ResData>>>' is not assignable to parameter of type 'Readonly<PublicConfiguration<any, any, BareFetcher<unknown>>>'.
  Types of property 'onLoadingSlow' are incompatible.
    Type '(key: string, config: Readonly<PublicConfiguration<ResData, ErrorResponse, (args: string) => FetcherResponse<ResData>>>) => void' is not assignable to type '(key: string, config: Readonly<PublicConfiguration<any, any, BareFetcher<unknown>>>) => void'.ts(2345)
```
I can fix it with the following by changing the `fetcher` type from the correct `(path: string) => Promise<ResData>` to `(path: string) => unknown` but it feels like a hack since it uses a wrong type and an import from `swr/_internal`.

```ts
import useSWR, { useSWRConfig } from 'swr';
import { PublicConfiguration } from 'swr/_internal';

  const { onErrorRetry } = useSWRConfig();

  return useSWR<ResData, ErrorResponse, string>(path, fetcher, {
    onErrorRetry(err, key, config, revalidate, revalidateOpts) {
      const configForDelegate = config as Readonly<PublicConfiguration<ResData, ErrorResponse, (path: string) => unknown>>;
      if (err.status === 404) return;
      onErrorRetry(err, key, configForDelegate, revalidate, revalidateOpts);
    },
  });
```
